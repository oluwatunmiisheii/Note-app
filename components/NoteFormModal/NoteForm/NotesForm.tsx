import { useState } from "react";
import { Text } from "react-native";
import { Note } from "@/models/note.model";
import { Button } from "@/components/Common/Button/Button";
import { Input } from "@/components/Common/Form/Input/Input";
import { useNote } from "@/hooks/useNote";

interface NotesFormProps {
  onSubmit?: (payload: Note) => void;
  isEditing?: boolean;
  defaultValues?: Note;
}

export const NotesForm = (props: NotesFormProps) => {
  const { onSubmit, isEditing, defaultValues } = props;
  const { createNote, updateNote } = useNote();
  const [inputs, setInputs] = useState({
    title: {
      value: defaultValues?.title ?? "",
      isValid: true,
    },
    description: {
      value: defaultValues?.description ?? "",
      isValid: true,
    },
    note: {
      value: defaultValues?.note ?? "",
      isValid: true,
    },
  });

  const handleInputChange = (value: string, name: string) => {
    setInputs((currentInputs) => ({
      ...currentInputs,
      [name]: {
        value,
        isValid: value.trim().length > 0,
      },
    }));
  };

  const validateFields = () => {
    const { title, description, note } = inputs;

    const titleIsValid = title.value.trim().length > 0;
    const descriptionIsValid = description.value.trim().length > 0;
    const noteIsValid = note.value.trim().length > 0;

    if (!titleIsValid || !descriptionIsValid || !noteIsValid) {
      setInputs((currentInputs) => {
        return {
          title: {
            value: currentInputs.title.value,
            isValid: titleIsValid,
          },
          description: {
            value: currentInputs.description.value,
            isValid: descriptionIsValid,
          },
          note: {
            value: currentInputs.note.value,
            isValid: noteIsValid,
          },
        };
      });
      return false;
    }
    return true;
  };

  const handleSubmit = () => {
    const { title, description, note } = inputs;

    const isValid = validateFields();

    if (!isValid) return;

    const payload = {
      title: title.value,
      description: description.value,
      note: note.value,
      date: defaultValues?.date ?? new Date(),
    };

    if (!isEditing) {
      createNote({
        payload,
        onSuccess: ({ name }) => {
          onSubmit?.({ ...payload, id: name });
        },
      });
    } else {
      updateNote({
        payload,
        noteId: defaultValues?.id ?? "",
        onSuccess: (note) => {
          onSubmit?.({
            ...note,
            id: defaultValues?.id ?? "",
            date: new Date(note.date),
          });
        },
      });
    }
  };

  return (
    <>
      <Input
        label="Title"
        onChangeText={(text) => handleInputChange(text, "title")}
        value={inputs.title.value}
        maxLength={50}
        testID="title-input"
        error={!inputs.title.isValid ? "Title is required" : undefined}
      />
      <Input
        label="Description"
        onChangeText={(text) => handleInputChange(text, "description")}
        value={inputs.description.value}
        multiline
        maxLength={200}
        testID="description-input"
        extraStyles={{ height: 100 }}
        error={
          !inputs.description.isValid ? "Description is required" : undefined
        }
      />
      <Input
        label="Note"
        onChangeText={(text) => handleInputChange(text, "note")}
        value={inputs.note.value}
        testID="note-input"
        extraStyles={{ height: "auto", minHeight: 250 }}
        error={!inputs.note.isValid ? "Note is required" : undefined}
        multiline
      />

      <Button
        onPress={handleSubmit}
        customStyles={{
          marginTop: 30,
        }}
        testID="submit-button"
      >
        <Text style={{ color: "white" }}>{isEditing ? "Update" : "Save"}</Text>
      </Button>
    </>
  );
};
