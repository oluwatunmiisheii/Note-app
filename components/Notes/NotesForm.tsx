import { useState } from "react";
import { Text } from "react-native";
import { useCreateOrUpdate } from "../../hooks/useCreateOrUpdate";
import { CreateNoteDto, Note } from "../../models/note.model";
import notesService from "../../services/note/notes.service";
import { Button } from "../common/Button/Button";
import { Input } from "../common/Form/Input/Input";

interface NotesFormProps {
  onSubmit?: (payload: Note) => void;
  isEditing?: boolean;
  defaultValues?: Note;
}

export const NotesForm = ({
  onSubmit,
  isEditing,
  defaultValues,
}: NotesFormProps) => {
  const createOrUpdate = useCreateOrUpdate();
  const [inputs, setInputs] = useState({
    title: {
      value: defaultValues?.title ?? "",
      isValid: true,
    },
    description: {
      value: defaultValues?.description ?? "",
      isValid: true,
    },
    note: {
      value: defaultValues?.note ?? "",
      isValid: true,
    },
  });

  const handleInputChange = (value: string, name: string) => {
    setInputs((currentInputs) => ({
      ...currentInputs,
      [name]: {
        value,
        isValid: value.trim().length > 0,
      },
    }));
  };

  const handleCreateNote = async (payload: CreateNoteDto) => {
    createOrUpdate({
      fn: notesService.createNote,
      args: [payload],
      successCb: (note) => {
        onSubmit?.({ ...payload, id: note.name });
      },
      successMessage: "Note created successfully",
      errorMessage: "Something went wrong",
    });
  };

  const handleUpdateNote = async (payload: CreateNoteDto) => {
    createOrUpdate({
      fn: notesService.updateNoteById,
      args: [defaultValues?.id ?? "", payload],
      successCb: (note) => {
        onSubmit?.({
          ...note,
          id: defaultValues?.id ?? "",
          date: new Date(note.date),
        });
      },
      successMessage: "Note updated successfully",
      errorMessage: "Something went wrong",
    });
  };

  const handleSubmit = () => {
    const { title, description, note } = inputs;

    const titleIsValid = title.value.trim().length > 0;
    const descriptionIsValid = description.value.trim().length > 0;
    const noteIsValid = note.value.trim().length > 0;

    if (!titleIsValid || !descriptionIsValid || !noteIsValid) {
      setInputs((currentInputs) => {
        return {
          title: {
            value: currentInputs.title.value,
            isValid: titleIsValid,
          },
          description: {
            value: currentInputs.description.value,
            isValid: descriptionIsValid,
          },
          note: {
            value: currentInputs.note.value,
            isValid: noteIsValid,
          },
        };
      });
      return;
    }

    const payload = {
      title: title.value,
      description: description.value,
      note: note.value,
      date: defaultValues?.date ?? new Date(),
    };

    isEditing ? handleUpdateNote(payload) : handleCreateNote(payload);
  };

  return (
    <>
      <Input
        label="Title"
        onChangeText={(text) => handleInputChange(text, "title")}
        value={inputs.title.value}
        maxLength={50}
        testID="title-input"
        error={!inputs.title.isValid ? "Title is required" : undefined}
      />
      <Input
        label="Description"
        onChangeText={(text) => handleInputChange(text, "description")}
        value={inputs.description.value}
        multiline
        maxLength={200}
        testID="description-input"
        extraStyles={{ height: 100 }}
        error={
          !inputs.description.isValid ? "Description is required" : undefined
        }
      />
      <Input
        label="Note"
        onChangeText={(text) => handleInputChange(text, "note")}
        value={inputs.note.value}
        testID="note-input"
        extraStyles={{ height: "auto", minHeight: 250 }}
        error={!inputs.note.isValid ? "Note is required" : undefined}
        multiline
      />

      <Button
        onPress={handleSubmit}
        customStyles={{
          marginTop: 30,
        }}
        testID="submit-button"
      >
        <Text style={{ color: "white" }}>{isEditing ? "Update" : "Save"}</Text>
      </Button>
    </>
  );
};
