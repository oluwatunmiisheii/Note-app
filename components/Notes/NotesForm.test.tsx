import renderer from "react-test-renderer";
import {
  render,
  fireEvent,
  screen,
  waitFor,
} from "@testing-library/react-native";
import notesService from "../../services/note/notes.service";

import { NotesForm } from "./NotesForm";
import { getNote } from "../../services/note/__mock_data__/notes.mock";

const mockOnSubmit = jest.fn();

const defaultProps: React.ComponentProps<typeof NotesForm> = {
  onSubmit: mockOnSubmit,
};

const create = (props: Partial<typeof defaultProps> = defaultProps) => {
  const tree = renderer
    .create(<NotesForm {...{ ...defaultProps, ...props }} />)
    .toJSON();
  return { tree };
};

const setup = (props: Partial<typeof defaultProps> = defaultProps) => {
  const utils = render(<NotesForm {...{ ...defaultProps, ...props }} />);
  return { ...utils };
};

afterEach(() => {
  jest.clearAllMocks();
});

describe(`<${NotesForm.name}>`, () => {
  test("renders correctly", () => {
    const { tree } = create();
    expect(tree).toMatchSnapshot();
  });

  describe("validation", () => {
    test("should validate title", () => {
      setup();
      const titleInput = screen.getByTestId("title-input");
      fireEvent.changeText(titleInput, "");
      fireEvent(titleInput, "blur");
      expect(titleInput.props.value).toBe("");

      expect(screen.getByText("Title is required")).toBeTruthy();
    });

    test("should validate description", () => {
      setup();
      const descriptionInput = screen.getByTestId("description-input");
      fireEvent.changeText(descriptionInput, "");
      fireEvent(descriptionInput, "blur");
      expect(descriptionInput.props.value).toBe("");

      expect(screen.getByText("Description is required")).toBeTruthy();
    });

    test("should validate note", () => {
      setup();
      const noteInput = screen.getByTestId("note-input");
      fireEvent.changeText(noteInput, "");
      fireEvent(noteInput, "blur");
      expect(noteInput.props.value).toBe("");

      expect(screen.getByText("Note is required")).toBeTruthy();
    });
  });

  describe("submit", () => {
    test("should call create note when isEditing is false or undefined", async () => {
      const createNoteSpy = jest
        .spyOn(notesService, "createNote")
        .mockImplementation(() => Promise.resolve({ name: "1" }));

      setup();

      getAndUpdateInput();

      const submitButton = screen.getByTestId("submit-button");
      fireEvent.press(submitButton);

      expect(createNoteSpy).toHaveBeenCalled();

      await waitFor(() => {
        expect(mockOnSubmit).toHaveBeenCalled();
      });
    });

    test("should call update note when isEditing is true", async () => {
      const updateNoteSpy = jest
        .spyOn(notesService, "updateNoteById")
        .mockImplementation(() =>
          Promise.resolve({
            name: "1",
            title: "title",
            description: "description",
            note: "note",
            date: new Date(),
          })
        );

      setup({ isEditing: true });

      getAndUpdateInput();

      const submitButton = screen.getByTestId("submit-button");
      fireEvent.press(submitButton);

      expect(updateNoteSpy).toHaveBeenCalled();

      await waitFor(() => {
        expect(mockOnSubmit).toHaveBeenCalled();
      });
    });
  });

  describe("default values are provided as props >", () => {
    test("should set default values", () => {
      const note = getNote();
      setup({ defaultValues: { ...note } });

      const titleInput = screen.getByTestId("title-input");
      const descriptionInput = screen.getByTestId("description-input");
      const noteInput = screen.getByTestId("note-input");

      expect(titleInput.props.value).toBe(note.title);
      expect(descriptionInput.props.value).toBe(note.description);
      expect(noteInput.props.value).toBe(note.note);
    });
  });
});

const getAndUpdateInput = () => {
  const titleInput = screen.getByTestId("title-input");
  const descriptionInput = screen.getByTestId("description-input");
  const noteInput = screen.getByTestId("note-input");

  fireEvent.changeText(titleInput, "title");
  fireEvent.changeText(descriptionInput, "description");
  fireEvent.changeText(noteInput, "note");
};
