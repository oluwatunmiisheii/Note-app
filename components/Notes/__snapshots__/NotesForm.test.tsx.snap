// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotesForm> renders correctly 1`] = `
Array [
  <View
    style={
      Object {
        "marginBottom": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#6c757d",
          "fontSize": 18,
          "fontWeight": "500",
          "marginBottom": 5,
        }
      }
    >
      Title
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#F5F5F5",
          "borderRadius": 10,
          "justifyContent": "space-between",
          "paddingHorizontal": 10,
          "paddingVertical": 5,
          "width": "100%",
        }
      }
    >
      <TextInput
        accessibilityLabel="input"
        label="Title"
        maxLength={50}
        onChangeText={[Function]}
        placeholderTextColor="#BDBDBD"
        style={
          Array [
            Array [
              Object {
                "fontSize": 18,
                "fontWeight": "700",
                "height": 50,
                "paddingHorizontal": 2,
                "textAlignVertical": "top",
              },
              undefined,
            ],
            undefined,
          ]
        }
        testID="title-input"
        value=""
      />
    </View>
  </View>,
  <View
    style={
      Object {
        "marginBottom": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#6c757d",
          "fontSize": 18,
          "fontWeight": "500",
          "marginBottom": 5,
        }
      }
    >
      Description
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#F5F5F5",
          "borderRadius": 10,
          "justifyContent": "space-between",
          "paddingHorizontal": 10,
          "paddingVertical": 5,
          "width": "100%",
        }
      }
    >
      <TextInput
        accessibilityLabel="input"
        extraStyles={
          Object {
            "height": 100,
          }
        }
        label="Description"
        maxLength={200}
        multiline={true}
        onChangeText={[Function]}
        placeholderTextColor="#BDBDBD"
        style={
          Array [
            Array [
              Object {
                "fontSize": 18,
                "fontWeight": "700",
                "height": 50,
                "paddingHorizontal": 2,
                "textAlignVertical": "top",
              },
              Object {
                "height": 250,
              },
            ],
            Object {
              "height": 100,
            },
          ]
        }
        testID="description-input"
        value=""
      />
    </View>
  </View>,
  <View
    style={
      Object {
        "marginBottom": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#6c757d",
          "fontSize": 18,
          "fontWeight": "500",
          "marginBottom": 5,
        }
      }
    >
      Note
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#F5F5F5",
          "borderRadius": 10,
          "justifyContent": "space-between",
          "paddingHorizontal": 10,
          "paddingVertical": 5,
          "width": "100%",
        }
      }
    >
      <TextInput
        accessibilityLabel="input"
        extraStyles={
          Object {
            "height": "auto",
            "minHeight": 250,
          }
        }
        label="Note"
        multiline={true}
        onChangeText={[Function]}
        placeholderTextColor="#BDBDBD"
        style={
          Array [
            Array [
              Object {
                "fontSize": 18,
                "fontWeight": "700",
                "height": 50,
                "paddingHorizontal": 2,
                "textAlignVertical": "top",
              },
              Object {
                "height": 250,
              },
            ],
            Object {
              "height": "auto",
              "minHeight": 250,
            },
          ]
        }
        testID="note-input"
        value=""
      />
    </View>
  </View>,
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 30,
            "height": 60,
            "justifyContent": "center",
            "padding": 5,
            "width": "auto",
          },
          Object {
            "backgroundColor": "#007bff",
          },
        ],
        undefined,
        Object {
          "marginTop": 30,
        },
      ]
    }
    testID="submit-button"
  >
    <Text
      style={
        Object {
          "color": "white",
        }
      }
    >
      Save
    </Text>
  </View>,
]
`;
